src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java

```diff
@@ -502,33 +502,41 @@ public static PqInfo getPqIfPresent(IndexContext indexContext, Function<VectorCo
     {
         // Retrieve the first compressed vectors for a segment with at least MAX_PQ_TRAINING_SET_SIZE rows
         // or the one with the most rows if none reach that size
-        var indexes = new ArrayList<>(indexContext.getView().getIndexes());
-        indexes.sort(Comparator.comparing(SSTableIndex::getSSTable, CompactionSSTable.maxTimestampDescending));
-
-        PqInfo cvi = null;
-        long maxRows = 0;
-        for (SSTableIndex index : indexes)
+        var view = indexContext.getReferencedView();
+        try
         {
-            for (Segment segment : index.getSegments())
-            {
-                if (segment.metadata.numRows < maxRows)
-                    continue;
+            var indexes = new ArrayList<>(view.getIndexes());
+            indexes.sort(Comparator.comparing(SSTableIndex::getSSTable, CompactionSSTable.maxTimestampDescending));
 
-                var searcher = (V2VectorIndexSearcher) segment.getIndexSearcher();
-                var cv = searcher.getCompression();
-                if (matcher.apply(cv))
+            PqInfo cvi = null;
+            long maxRows = 0;
+            for (SSTableIndex index : indexes)
+            {
+                for (Segment segment : index.getSegments())
                 {
-                    // We can exit now because we won't find a better candidate
-                    var candidate = new PqInfo(searcher.getPQ(), searcher.containsUnitVectors(), segment.metadata.numRows);
-                    if (segment.metadata.numRows >= ProductQuantization.MAX_PQ_TRAINING_SET_SIZE)
-                        return candidate;
-
-                    cvi = candidate;
-                    maxRows = segment.metadata.numRows;
+                    if (segment.metadata.numRows < maxRows)
+                        continue;
+
+                    var searcher = (V2VectorIndexSearcher) segment.getIndexSearcher();
+                    var cv = searcher.getCompression();
+                    if (matcher.apply(cv))
+                    {
+                        // We can exit now because we won't find a better candidate
+                        var candidate = new PqInfo(searcher.getPQ(), searcher.containsUnitVectors(), segment.metadata.numRows);
```

Comment:
TODO: make sure the PQ is safe to use after releasing the view.

Reply:
how about adding a `SSTableIndex` in `PqInfo` and release it when `PqInfo` is done